src/libraries/Native/Unix/System.Native/pal_random.c

```diff
@@ -20,91 +21,234 @@ Generate random bytes. The generated bytes are not cryptographically strong.
+    // TryAllRandomFillBufferMethods() will be either all zeroed out or
+    // will have all bits set.  This happens in some rare occasions in
+    // some systems (e.g.  resuming from suspend).  Try to detect this
+    // situation here and XOR every byte with something obtained from
+    // srand48().
+
+    if (bufferLength >= sizeof(uint32_t))
     {
-        srand48(time(NULL));
-        sInitializedMRand = true;
+        uint32_t first32bits;
+
+        memcpy(&first32bits, buffer, sizeof(first32bits));
+
+        return (first32bits == 0u) || (first32bits == 0xffffffffu);
     }
 
-    // always xor srand48 over the whole buffer to get some randomness
-    // in case /dev/urandom is not really random
+    return !memcmp(buffer, (const uint8_t[]) {0x00, 0x00, 0x00}, bufferLength) ||
+           !memcmp(buffer, (const uint8_t[]) {0xff, 0xff, 0xff}, bufferLength);
+}
+
+void SystemNative_GetNonCryptographicallySecureRandomBytes(uint8_t *buffer, int32_t bufferLengthAsInt)
+{
+    size_t bufferLength = (size_t)bufferLengthAsInt;
+    bool filledBuffers = TryAllRandomFillBufferMethods(buffer, bufferLength);
 
-    for (int i = 0; i < bufferLength; i++)
+    if (MAY_NEED_XOR_WITH_LRAND && (!filledBuffers || NeedsXorWithLrand(buffer, bufferLength)))
     {
-        if (i % 4 == 0)
+        static bool sInitializedMRand;
+
+        if (!sInitializedMRand)
+        {
+            srand48(time(NULL));
+            sInitializedMRand = true;
+        }
+
+        if (!filledBuffers)
         {
-            num = lrand48();
+            // Zero out buffer first to avoid UB when reading back from buffer
+            memset(buffer, 0, bufferLength);
         }
 
-        *(buffer + i) ^= num;
-        num >>= 8;
+        while (bufferLength >= sizeof(uint32_t))
+        {
+            uint32_t asInteger;
+
+            memcpy(&asInteger, buffer, sizeof(asInteger));
+            asInteger ^= (uint32_t)lrand48();
+            memcpy(buffer, &asInteger, sizeof(asInteger));
+
+            buffer += sizeof(asInteger);
+            bufferLength -= sizeof(asInteger);
+        }
+
+        if (bufferLength)
+        {
+            long int lastLrandRound = lrand48();
+
+            switch (bufferLength)
+            {
+                case 3:
+                    buffer[2] ^= lastLrandRound & 0xff;
+                    lastLrandRound >>= 8;
+                    // Fallthrough
+                case 2:
+                    buffer[1] ^= lastLrandRound & 0xff;
+                    lastLrandRound >>= 8;
+                    // Fallthrough
+                case 1:
+                    buffer[0] ^= lastLrandRound & 0xff;
+                    break;
+                default:
+                    assert(bufferLength < sizeof(uint32_t));
+            }
+
+            lastLrandRound = 0;
+            // Ensure dead store elimination does not yank the previous assignment
+            __asm__ __volatile__("" ::"g"(lastLrandRound) : "memory");
```

Comment:
I am not sure I understand why we assign to the variable right before it goes out of scope at all.

Reply:
It's a common defensive programming strategy to avoid having potentially sensitive data in some register.  It doesn't guarantee much (e.g. vectorized `memcpy()` can keep stuff in SIMD registers), though.

Reply:
Why is it important to use this defensive strategy here?